### PR TITLE
Improve notifications popup

### PR DESCRIPTION
**Pull request by Aivean**
### Changes summary:
* separate read/hidden status with a separate "delete" button
* read notifications remain in the view until manually deleted
* styles
  * overlay in the background
  * max height 100%
  * read/unread

image

image

Delete button hover:
image

### CODE REVIEW

File: frontend/src/Components/NotificationsPopup.tsx [hunk 1]
```diff
@@ -76,50 +79,80 @@ export default function NotificationsPopup(props: NotificationsPopupProps) {
...
-            </div>
-            <div className={styles.buttons}>
-                <button className={styles.buttonClear} onClick={handleClearAll}>Очистить</button>
-                <button className={styles.buttonAll} disabled={true}>Все чпяки</button>
+        <>
+            <div className={styles.overlay} onClick={props.onClose}/>
+            <div className={styles.container}>
+                <div className={styles.notifications}>
+                    {error}
+                    {notifications && notifications.map(notify => {
```
Comment (4vanger) [inline frontend/src/Components/NotificationsPopup.tsx:131]: nit
```suggestion
                    {notifications && notifications.map(notification=> {
```

File: frontend/src/Components/NotificationsPopup.tsx [hunk 2]
```diff
@@ -76,50 +79,80 @@ export default function NotificationsPopup(props: NotificationsPopupProps) {
...
+                            key={notify.id}
+                            className={classNames(styles.notification, {[styles.read]: notify.read})}
+                            post={{ id: notify.source.post.id, site: notify.source.post.site }}
+                            commentId={notify.source.comment?.id}
+                            onClick={handleNotificationClick(notify)}
+                            onlyNew={true}
+                        >
+                            <div className={styles.type}>{notify.type === 'answer' ? <CommentIcon /> : <MentionIcon />}</div>
+                            <div className={styles.content}>
+                                <div className={styles.date}>{notify.source.byUser.username} {notify.type === 'answer' ? 'ответил вам' : 'упомянул вас'} <DateComponent date={notify.date} /></div>
```
Comment (4vanger) [inline frontend/src/Components/NotificationsPopup.tsx:142]: Надо бы сделать это gender neutral. Ток чот голова не варит как это сделать на русском. 'ответил(а)/написал(а)'?

File: frontend/src/Components/NotificationsPopup.tsx [hunk 3]
```diff
@@ -76,50 +79,80 @@ export default function NotificationsPopup(props: NotificationsPopupProps) {
...
+                                <div className={styles.text}>{notify.source.comment?.content}</div>
+                            </div>
+                            <div className={styles.remove} onClick={handleNotificationHide(notify)}><CloseIcon/></div>
+                        </PostLink>;
+                    })}
+                </div>
+                <div className={styles.buttons}>
+                    <button className={styles.buttonClear} onClick={handleClearAll}>Очистить</button>
+                    <button className={styles.buttonRead} onClick={handleReadAll}>Прочитать все</button>
+                    {/*<button className={styles.buttonAll} disabled={true}>Все чпяки</button>*/}
```
Comment (4vanger) [inline frontend/src/Components/NotificationsPopup.tsx:152]: remove

File: frontend/src/Components/Topbar.tsx [hunk 4]
```diff
@@ -89,8 +89,15 @@ const WatchButton = observer(() => {
 });
 
 const NotificationsButton = observer((props: React.ComponentPropsWithRef<'button'>) => {
-    const {notificationsCount} = useAppState();
+    const {unreadNotificationsCount, visibleNotificationsCount} = useAppState();
     return (
-        <button {...props} disabled={notificationsCount === 0} className={notificationsCount > 0 ? styles.active : ''}><NotificationIcon /><span className={styles.label}>{notificationsCount > 0 ? notificationsCount : ''}</span></button>
+        <button {...props} disabled={visibleNotificationsCount === 0}
+                className={unreadNotificationsCount > 0 ? styles.active : ''}><NotificationIcon />
```
Comment (4vanger) [inline frontend/src/Components/Topbar.tsx:95]: Maybe a slightly more readable
```suggestion
                className={classNames({[styles.active]: unreadNotificationsCount})><NotificationIcon />
```

File: frontend/src/Components/Topbar.tsx [hunk 5]
```diff
@@ -89,8 +89,15 @@ const WatchButton = observer(() => {
...
+            <span className={styles.label}>
+                {visibleNotificationsCount > 0 ?
```
Comment (4vanger) [inline frontend/src/Components/Topbar.tsx:97]: two ternary if statements are really hard to read.
unreadNotificationsCount !== visibleNotificationsCount ? `${unreadNotificationsCount}/${visibleNotificationsCount}` : (visibleNotificationsCount || '')

File: frontend/src/Components/NotificationsPopup.tsx [hunk 6]
```diff
@@ -76,50 +79,80 @@ export default function NotificationsPopup(props: NotificationsPopupProps) {
...
         if (props.onClose) {
             props.onClose();
         }
     };
 
+    const handleReadAll = () => {
+        app.setUnreadNotificationsCount(0);
+        api.notifications.readAll()
+            .then()
+            .catch();
```
Comment (4vanger) [inline frontend/src/Components/NotificationsPopup.tsx:122]: Output error through `console.error` maybe? Silently swallowing error will make it really hard to debug.

File: frontend/src/Components/NotificationsPopup.tsx [hunk 7]
```diff
@@ -76,50 +79,80 @@ export default function NotificationsPopup(props: NotificationsPopupProps) {
...
+        setNotifications(notifications?.filter(n => n.id !== notify.id));
+    };
+
     const handleClearAll = () => {
-        api.notifications.readAll()
+        app.setVisibleNotificationsCount(0);
+        app.setUnreadNotificationsCount(0);
+        api.notifications.hideAll()
             .then()
             .catch();
```
Comment (4vanger) [inline frontend/src/Components/NotificationsPopup.tsx:112]: Output error through `console.error` maybe? Silently swallowing error will make it really hard to debug.

File: frontend/src/Components/NotificationsPopup.tsx [hunk 8]
```diff
@@ -76,50 +79,80 @@ export default function NotificationsPopup(props: NotificationsPopupProps) {
...
             props.onClose();
         }
     };
 
+    const handleNotificationHide = (notify: NotificationInfo) => (e: React.MouseEvent<unknown>) => {
+        e.preventDefault();
+        e.stopPropagation();
+        api.notifications.hide(notify.id)
+            .then()
+            .catch();
```
Comment (4vanger) [inline frontend/src/Components/NotificationsPopup.tsx:99]: Output error through `console.error` maybe? Silently swallowing error will make it really hard to debug.

File: frontend/src/Components/NotificationsPopup.tsx [hunk 9]
```diff
@@ -76,50 +79,80 @@ export default function NotificationsPopup(props: NotificationsPopupProps) {
             });
     }, [fetchNotifications, subscribe]);
 
-    const handleNotificationClick = (e: React.MouseEvent<HTMLAnchorElement>, notify: NotificationInfo) => {
+    const handleNotificationClick = (notify: NotificationInfo) => (e: React.MouseEvent<HTMLAnchorElement>) => {
         api.notifications.read(notify.id)
             .then()
             .catch();
```
Comment (4vanger) [inline frontend/src/Components/NotificationsPopup.tsx:85]: Output error through `console.error` maybe? Silently swallowing error will make it really hard to debug.

**Review by 4vanger — APPROVED** [review]
LGTM. Few minor notes

File: frontend/src/Components/NotificationsPopup.tsx (hunk 2 above)
Comment (Aivean) [reply frontend/src/Components/NotificationsPopup.tsx:142]: We already have the ability to specify user gender. Will try to use it here.

File: frontend/src/Components/NotificationsPopup.tsx (hunk 3 above)
Comment (Aivean) [reply frontend/src/Components/NotificationsPopup.tsx:152]: Code and styles are still useful. Will mark as TODO.

File: frontend/src/Components/Topbar.tsx (hunk 5 above)
Comment (Aivean) [reply frontend/src/Components/Topbar.tsx:97]: That's not the same condition. I extracted the condition into explicit `if` statements. Hope it's more readable now.

File: frontend/src/Components/NotificationsPopup.tsx (hunk 8 above)
Comment (Aivean) [reply frontend/src/Components/NotificationsPopup.tsx:99]: Added toast.

File: frontend/src/Components/NotificationsPopup.tsx (hunk 2 above)
Comment (Aivean) [reply frontend/src/Components/NotificationsPopup.tsx:142]: Fixed:
image

**Comment by Aivean** [comment]
> LGTM. Few minor notes

Thanks for the review. Addressed review comments.

File: frontend/src/Components/Topbar.tsx [hunk 10]
```diff
@@ -90,14 +91,19 @@ const WatchButton = observer(() => {
 
 const NotificationsButton = observer((props: React.ComponentPropsWithRef<'button'>) => {
     const {unreadNotificationsCount, visibleNotificationsCount} = useAppState();
+
+    let label = '';
```
Comment (4vanger) [inline frontend/src/Components/Topbar.tsx:95]: :heart_eyes: :heart_eyes: :heart_eyes:

**Review by 4vanger — APPROVED** [review]

**Pull request merged**